### ISSUE TITLE
Fix for Mojave Pls [Error due to Mismatched Kernel Headers built against]

### ISSUE THREAD

**Issue by Genxster1998** [zxystd/AthBluetoothFirmware#3]
```The super class vtable '__ZTV9IOService' for vtable '__ZTV16AtherosFWService' is out of date. Make sure your kext has been built against the correct headers.```

Above error on my 10.14.x is absolute result of  kext built against new kernel headers ,I have seen similar errors on a lot of kexts .

**there is some  serious changes in handling usb iokit in 10.15+ ?** 

```Update (Needs to Rewrite Driver)```

Reason :
[USBHOST DRIVER Requirements 10.15+](https://developer.apple.com/documentation/iousbhost)
 
~~Will you help me to get it done with legacy usb iokit in ar3011 ath3kfirmware ?
Ar3011 doesn't need a thing other than injecting to endpoint 2 while this requires setting device and releasing /changing vid/pid etc.~~

**Comment by Genxster1998** [comment]
So Today I compiled it(ath3kBT only) against macossdk10.14  but no gain instantly result  in kp on loading. Injector Loads fines already .
```
Fri Mar 26 02:48:52 2021

*** Panic Report ***
panic(cpu 2 caller 0xffffff8011edbadd): Kernel trap at 0xffffff8012422a90, type 14=page fault, registers:
CR0: 0x000000008001003b, CR2: 0x0000000000000000, CR3: 0x000000001bb41000, CR4: 0x00000000001626e0
RAX: 0x0000000020000000, RBX: 0x0000000000000000, RCX: 0x0000000020000001, RDX: 0xffffff802372d000
RSP: 0xffffff80b1143e50, RBP: 0xffffff80b1143e60, RSI: 0xffffff8022e3b760, RDI: 0x0000000000000000
R8:  0xffffff802372d000, R9:  0xffffff80126a3ea0, R10: 0xffffff80126a3ec0, R11: 0xffffff8022e6bd60
R12: 0xffffff802372d000, R13: 0xffffff8023772600, R14: 0xffffff802372d000, R15: 0xffffff802324e360
RFL: 0x0000000000010246, RIP: 0xffffff8012422a90, CS:  0x0000000000000008, SS:  0x0000000000000000
Fault CR2: 0x0000000000000000, Error code: 0x0000000000000000, Fault CPU: 0x2, PL: 0, VF: 1

Backtrace (CPU 2), Frame : Return Address
0xffffff80b1143920 : 0xffffff8011dae6ed 
0xffffff80b1143970 : 0xffffff8011eea185 
0xffffff80b11439b0 : 0xffffff8011edb8ba 
0xffffff80b1143a20 : 0xffffff8011d5bb40 
0xffffff80b1143a40 : 0xffffff8011dae107 
0xffffff80b1143b60 : 0xffffff8011dadf53 
0xffffff80b1143bd0 : 0xffffff8011edbadd 
0xffffff80b1143d40 : 0xffffff8011d5bb40 
0xffffff80b1143d60 : 0xffffff8012422a90 
0xffffff80b1143e60 : 0xffffff7f94c3a792 
0xffffff80b1143e80 : 0xffffff801242f884 
0xffffff80b1143f00 : 0xffffff801242f337 
0xffffff80b1143f50 : 0xffffff8012430e26 
0xffffff80b1143fa0 : 0xffffff8011d5b0ce 
      Kernel Extensions in backtrace:
         com.zxystd.Ath3kBT(1.0)[6F9DDC6E-3B49-345F-8650-AB7D17591DC3]@0xffffff7f94c3a000->0xffffff7f94c93fff
            dependency: com.apple.iokit.IOUSBHostFamily(1.2)[BD06015A-414D-34DF-BE36-7B9034D7360D]@0xffffff7f92728000

BSD process name corresponding to current thread: kernel_task
Boot args: alcid=3 vm_compressor=2 -v

Mac OS version:
Not yet set

Kernel version:
Darwin Kernel Version 18.7.0: Tue Aug 20 16:57:14 PDT 2019; root:xnu-4903.271.2~2/RELEASE_X86_64
Kernel UUID: C41337A1-0EC3-3896-A954-A1F675656
Kernel slide:     0x0000000011a00000
Kernel text base: 0xffffff8011c00000
__HIB  text base: 0xffffff8011b00000

System uptime in nanoseconds: 4458155973
```

**Comment by zxystd** [comment]
@Genxster1998 Nothing need to be change in source code, it is compatible to 10.12 and above, you just need to compile it with the correct SDK. BTW, what's the version of mac you are using to compile and caused the KP?

**Comment by Genxster1998** [comment]
@zxystd I am currently on 10.14.6 .Other user lynxnu said that he has tested it with 30+ users ,no body succeeded on Mojave or below .
I am currently using 10.14 as base sdk with MacKernelSDK .But it results in panic .
I see through your code , it's **fabulously totally solid implementation of linux ath3kc.** 
My Takes :-
- [x] 
I tried modifying  Rehabman AR3011 OSX-Ath3k-Firmware .
I can get state,version and other usb req  quirks work but sending buffer throws error and lead to bizzare condition of bluetooth device.

- [x] 
I ported libusb based bsd userland ath3kfw tool ,but sadly it can't claim interface 0 as already someone has .I tried fake kext but still IOUSBHostDevice doesn't want to leave it.

I am guessing how vmware likely get usb ownership transferred but userland tool even if get successful it won't be dynamic (sleep etc .).

**Comment by Genxster1998** [comment]
I will try other SDKs like 10.12 and 10.13

**Comment by zxystd** [comment]
[Ath3kBT.kext.zip](https://github.com/zxystd/AthBluetoothFirmware/files/6254025/Ath3kBT.kext.zip)
compile with 10.12 sdk on my test machine. @Genxster1998

**Comment by zxystd** [comment]
@Genxster1998  macOS USB IOKit stack changing a lot from 10.11, after that, they are almost the same in kernel API side, I think you get this loading error is because the project configuration mistakes somewhere, not about the code.

**Comment by Genxster1998** [comment]
> @Genxster1998 macOS USB IOKit stack changing a lot from 10.11, after that, they are almost the same in kernel API side, I think you get this loading error is because the project configuration mistakes somewhere, not about the code. BTW, I don't have ar3k devices to test, this repo is written by me and tested pass by others who have AR9565 chip, it works fine.

Same Error : Kernel Panic On Boot
`Sun Apr  4 20:56:20 2021

*** Panic Report ***
panic(cpu 1 caller 0xffffff801a8dbadd): Kernel trap at 0xffffff801ae22a90, type 14=page fault, registers:
CR0: 0x000000008001003b, CR2: 0x0000000000000000, CR3: 0x00000000246c2000, CR4: 0x00000000001626e0
RAX: 0x0000000020000000, RBX: 0x0000000000000000, RCX: 0x0000000020000001, RDX: 0xffffff7f9d9a3220
RSP: 0xffffff80b9cabe40, RBP: 0xffffff80b9cabe50, RSI: 0xffffff802bbfe7a0, RDI: 0x0000000000000000
R8:  0xffffff802c4f83c0, R9:  0x00000000000001de, R10: 0xffffff801b0a3ec0, R11: 0xffffff802bc31140
R12: 0xffffff802c4f83c0, R13: 0xffffff802c566e40, R14: 0xffffff802c4f83c0, R15: 0xffffff802c0170c0
RFL: 0x0000000000010246, RIP: 0xffffff801ae22a90, CS:  0x0000000000000008, SS:  0x0000000000000000
Fault CR2: 0x0000000000000000, Error code: 0x0000000000000000, Fault CPU: 0x1, PL: 0, VF: 1

Backtrace (CPU 1), Frame : Return Address
0xffffff80b9cab910 : 0xffffff801a7ae6ed 
0xffffff80b9cab960 : 0xffffff801a8ea185 
0xffffff80b9cab9a0 : 0xffffff801a8db8ba 
0xffffff80b9caba10 : 0xffffff801a75bb40 
0xffffff80b9caba30 : 0xffffff801a7ae107 
0xffffff80b9cabb50 : 0xffffff801a7adf53 
0xffffff80b9cabbc0 : 0xffffff801a8dbadd 
0xffffff80b9cabd30 : 0xffffff801a75bb40 
0xffffff80b9cabd50 : 0xffffff801ae22a90 
0xffffff80b9cabe50 : 0xffffff7f9d94a8fb 
0xffffff80b9cabe80 : 0xffffff801ae2f884 
0xffffff80b9cabf00 : 0xffffff801ae2f337 
0xffffff80b9cabf50 : 0xffffff801ae30e26 
0xffffff80b9cabfa0 : 0xffffff801a75b0ce 
      Kernel Extensions in backtrace:
         com.zxystd.Ath3kBT(1.0)[1C4F9F5D-BEED-36DA-BBA0-620E6626804E]@0xffffff7f9d94a000->0xffffff7f9d9a4fff
            dependency: com.apple.iokit.IOUSBHostFamily(1.2)[BD06015A-414D-34DF-BE36-7B9034D7360D]@0xffffff7f9b2ad000

BSD process name corresponding to current thread: kernel_task
Boot args: alcid=3 vm_compressor=2

Mac OS version:
Not yet set

Kernel version:
Darwin Kernel Version 18.7.0: Tue Aug 20 16:57:14 PDT 2019; root:xnu-4903.271.2~2/RELEASE_X86_64
Kernel UUID: C41337A1-0EC3-3896-A954-A1F85E849D53
Kernel slide:     0x000000001a400000
Kernel text base: 0xffffff801a600000
__HIB  text base: 0xffffff801a500000
System model name: MacBookPro11,1 (Mac-189A3D4F975D5FFC)

System uptime in nanoseconds: 3827188066

`

**Comment by Genxster1998** [comment]
@zxystd Problem is most devices with AR9565 adapter are old and i dont know how all your testers happen to be  on Catelina+ which is more resource demanding than previous versions.
I will be obliged if you can let me see any one with 10.12-10.14 running this kext.Thanks For your Prompt Reply.

**Comment by zxystd** [comment]
[Ath3kBT.kext.zip](https://github.com/zxystd/AthBluetoothFirmware/files/6255772/Ath3kBT.kext.zip)
@Genxster1998 I found an API behavior changed, please Try this one again.

**Comment by Genxster1998** [comment]
@zxystd okay I will try again just now

**Comment by Genxster1998** [comment]
@zxystd WoooH! Its Fixed .
```
ocalhost kernel[0]: (Ath3kBT) Ath3kBT::probe
2021-04-05 07:04:58.166509+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT::probe
2021-04-05 07:04:58.173544+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT::start!
2021-04-05 07:04:58.180272+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT::message
2021-04-05 07:04:58.186809+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT::start: device reset
2021-04-05 07:04:58.193656+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT::start: num configurations 1
2021-04-05 07:04:58.202250+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT::message
2021-04-05 07:04:58.208436+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT::message
2021-04-05 07:04:58.214628+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT::message
2021-04-05 07:04:58.220651+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT::message
2021-04-05 07:04:58.226862+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT::message
2021-04-05 07:04:58.233100+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT::message
2021-04-05 07:04:58.238779+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT::start Found Bulk out endpoint!
2021-04-05 07:04:58.245970+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT try to load patch rom file AthrBT_0x31010000.dfu
2021-04-05 07:04:58.305477+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT try to load syscfg file ramps_0x31010000_40.dfu
2021-04-05 07:04:58.314216+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT firmware was already in normal mode
2021-04-05 07:04:58.318405+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT: firmware loaded successfully!
2021-04-05 07:04:58.319098+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT::start: device status 3
2021-04-05 07:04:58.326628+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT::message
2021-04-05 07:05:01.722502+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT::message
2021-04-05 07:05:01.733098+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT::message
2021-04-05 07:05:01.733460+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT::message
2021-04-05 07:05:01.735147+0530  localhost kernel[0]: (Ath3kBT) Ath3kBT::message
```
it was swift as normal boot .Please reflect changes in repo .
Thanks

**Comment by zxystd** [comment]
> it was swift as normal boot .Please reflect changes in repo .

Ok, I will commit and push soon. Thank you for pointed out this problem and help to test.

> Thanks ,BTW do you happen to know any fix for Handsoff and Continuity like for Broadcomm bluetooth adapter its not enabled here.

Of course, the WiFi driver should report the AWDL/HandOff capability flags correctly to Airport Framework which is different with older flags in AtherosAirport driver. So, we should write a patch kext to enable them(Hardcoded in kext file is also a choice).
for example, that's what I do in Intel WiFi driver:

https://github.com/OpenIntelWireless/itlwm/tree/master/AirportItlwm
```
IOReturn AirportItlwm::
getCARD_CAPABILITIES(OSObject *object,
                                     struct apple80211_capability_data *cd)
{
    uint32_t caps = fHalService->get80211Controller()->ic_caps;
    if (caps & IEEE80211_C_WEP)
        cd->capabilities[0] |= 1 << APPLE80211_CAP_WEP;
    if (caps & IEEE80211_C_RSN)
        cd->capabilities[0] |= 1 << APPLE80211_CAP_TKIP | 1 << APPLE80211_CAP_AES_CCM;
    // Disable not implemented capabilities
    // if (caps & IEEE80211_C_PMGT)
    //     cd->capabilities[0] |= 1 << APPLE80211_CAP_PMGT;
    // if (caps & IEEE80211_C_IBSS)
    //     cd->capabilities[0] |= 1 << APPLE80211_CAP_IBSS;
    // if (caps & IEEE80211_C_HOSTAP)
    //     cd->capabilities[0] |= 1 << APPLE80211_CAP_HOSTAP;
    // AES not enabled, like on Apple cards
    
    if (caps & IEEE80211_C_SHSLOT)
        cd->capabilities[1] |= 1 << (APPLE80211_CAP_SHSLOT - 8);
    if (caps & IEEE80211_C_SHPREAMBLE)
        cd->capabilities[1] |= 1 << (APPLE80211_CAP_SHPREAMBLE - 8);
    if (caps & IEEE80211_C_RSN)
        cd->capabilities[1] |= 1 << (APPLE80211_CAP_WPA1 - 8) | 1 << (APPLE80211_CAP_WPA2 - 8) | 1 << (APPLE80211_CAP_TKIPMIC - 8);
    // Disable not implemented capabilities
    // if (caps & IEEE80211_C_TXPMGT)
    //     cd->capabilities[1] |= 1 << (APPLE80211_CAP_TXPMGT - 8);
    // if (caps & IEEE80211_C_MONITOR)
    //     cd->capabilities[1] |= 1 << (APPLE80211_CAP_MONITOR - 8);
    // WPA not enabled, like on Apple cards

    cd->version = APPLE80211_VERSION;
    cd->capabilities[2] = 0xFF; // BURST, WME, SHORT_GI_40MHZ, SHORT_GI_20MHZ, WOW, TSN, ?, ?
    cd->capabilities[3] = 0x2B;
    cd->capabilities[4] = 0xAD;
    cd->capabilities[5] = 0x80;//isCntryDefaultSupported
    cd->capabilities[5] |= 0x0C;
    cd->capabilities[6] = 0x8D;
    cd->capabilities[7] = 0x84; // This byte contains Apple Watch unlock
    //cd->capabilities[8] = 0x40;
    //cd->capabilities[8] |= 8;//dfs white list
    //cd->capabilities[9] = 0x28;
    return kIOReturnSuccess;
}
```

**Comment by Genxster1998** [comment]
@zxystd you mean to say it has to be another patch kext for wifi adapter or bluetooth adapter ? Isn't personification as Broadcomm Adapter enough ? Do you happen to have  fix for ar9565 wifi tx rate limit to 11mbps ?

**Comment by zxystd** [comment]
@Genxster1998 A patch kext like this one.
https://github.com/acidanthera/AirportBrcmFixup

I don't sure if it will work, because Atheros AWDL implementation isn't the same as Broadcom, hmmm, Atheros use out of date AWDL protocol version in brief.

I don't know the rate limit, I have ar9565 in my old-old laptop and don't use it anymore.

**Comment by Genxster1998** [comment]
@zxystd Thanks for explaining details .
You can close this.

**Comment by zxystd** [comment]
fix in 126fd237b2d280ac95889f8499d6a00135c08f6c
